crawl-ref/source/catch2-tests/test_fixture_lua.cc

```diff
@@ -0,0 +1,283 @@
+/// \file
+/// \brief Test some features of the lua fixture
+/// \details
+/// - Setup and tear down can happen for more than once
+/// - Monster placement (dgn.create_monster)
+/// - you.mutate, moveto
+/// - tile and terrain
+
+#include "catch.hpp"
+#include "AppHdr.h"
+#include "fixture_lua.h"
+#include "stringutil.h"
+#include "rltiles/tiledef-dngn.h"
+
+// -------------------- //
+// Forward declarations //
+// -------------------- //
+void place_monster(const coord_def& pos, const std::string& spec);
+
+void require_mons_empty(const int);
+
+// ------------- //
+// General tests //
+// ------------- //
+
+/// \brief Setup and tear down the fixture for 3 times
+TEST_CASE("fixture_lua: Test setup and teardown",
+          "[single-file][test_fixture_lua]")
+{
+    for (int i = 0; i < 5; ++i)
+    {
+        details_fixture_lua::_setup_fixture_lua();
+        details_fixture_lua::_teardown_fixture_lua();
+    }
+}
+
+/// \brief Setup and tear down the fixture for 3 times
+TEST_CASE("fixture_lua: Test running a Lua command",
+          "[single-file][test_fixture_lua]")
+{
+    for (int i = 0; i < 5; ++i)
+    {
+        details_fixture_lua::_setup_fixture_lua();
+        require_execstring("return 1 + 1", 1);
+        details_fixture_lua::_teardown_fixture_lua();
+    }
+}
+
+/// \brief Test running a Lua script
+TEST_CASE("fixture_lua: Test running a Lua script",
+          "[single-file][test_fixture_lua]")
+{
+    for (int i = 0; i < 5; ++i)
+    {
+        details_fixture_lua::_setup_fixture_lua();
+        you.hp = 100;  // need this or it crashes
+        dlua.execfile("mutation.lua");
+        details_fixture_lua::_teardown_fixture_lua();
+    }
+}
```

Comment:
These tests check if the setup and tear-down of the test can successfully reset part of the crawl's game state.

The test function by setup, check, tear down, setup, check again, and tear down. The 2nd check happens after the 1st tear-down to see if the tear-down can reset the part of the game state.

Directly using the fixture in the tests means I get setup, check, and tear-down. There is no tear-down before the check and I can't figure out if the setup and tear-down together reset the part of the game state.

This is the only reason that I expose:

-  `void _setup_fixture_lua()` and 
- `void _teardown_fixture_lua()`

Reply:
I think I can use the fixture for these 3 particular tests. I haven't figured out how to use the fixture directly in the other tests because the other tests need multiple setup and tear-down steps.

Maybe the the following can work?
```c++
TEST_CASE("xyz")
{
    for (int i = 0; i < 5; ++i)
    {
        Fixture_lua(); // this goes out of scope in each iteration.
        // checks
    }
}
```

Reply:
Should be `Fixture_lua fl;` instead of `Fixture_lua()`. Otherwise, the `Fixture_lua` object would be immediately destroyed after its creation.